javalib/src/main/scala/java/io/InputStream.scala

```diff
@@ -44,6 +46,68 @@ abstract class InputStream extends Closeable {
     }
   }
 
+  def readAllBytes(): Array[Byte] =
+    readNBytes(Integer.MAX_VALUE)
+
+  def readNBytes(len: Int): Array[Byte] = {
+    if (len < 0) {
+      throw new IllegalArgumentException
+    } else if (len == 0) {
+      new Array[Byte](0)
+    } else {
+      var bytesRead = 0
+
+      /* Allocate a buffer.
+       *
+       * Note that the implementation is required to grow memory proportional to
+       * the amount read, not the amount requested. Therefore, we cannot simply
+       * allocate an array of length len.
+       */
+      var buf = new Array[Byte](1024)
```

Comment:
Perhaps allocate exactly `len` bytes if it is smaller than 1024. This way, we avoid the copy at the end. We can also take advantage of `available()` to allocate the full buffer from the start if we know there will be enough bytes:
```scala
val initialBufLength = if (len <= 1024 || available() >= len) len else 1024
```

Reply:
Hmmm... It is even tempting to allocate `Math.min(len, Math.max(available(), 1024))`.

However, `available()` doesn't actually give us a guarantee about how many bytes will be read. Only for how many we can ask without blocking. So technically, an implementation returning Integer.MAX_VALUE whenever one byte is available is correct.

OTOH, it might be more reasonable to still do this.

I'll go with `Math.min(len, 1024)` for now.